src/codec/framers/newline_delimited.rs

```diff
@@ -0,0 +1,142 @@
+    pub fn new() -> Self {
+        Self { max_length: None }
+    }
+}
+
+#[typetag::serde(name = "newline_delimited")]
+impl FramingConfig for NewlineDelimitedDecoderConfig {
+    fn build(&self) -> BoxedFramer {
+        Box::new(
+            self.max_length
+                .map(|max_length| NewlineDelimitedCodec::new_with_max_length(max_length))
+                .unwrap_or_else(|| NewlineDelimitedCodec::new()),
+        )
+    }
+}
+
+/// A codec for handling bytes that are delimited by (a) newline(s).
+#[derive(Debug, Clone)]
+pub struct NewlineDelimitedCodec(LinesCodec);
+
+impl NewlineDelimitedCodec {
+    /// Creates a new `NewlineDelimitedCodec`.
+    pub fn new() -> Self {
+        Self(LinesCodec::new())
+    }
+
+    /// Creates a new `NewlineDelimitedCodec` with `max_length`.
+    ///
+    /// When more bytes than `max_length` have been read, all bytes will be
+    /// discarded until reaching the next newline.
+    pub fn new_with_max_length(max_length: usize) -> Self {
+        Self(LinesCodec::new_with_max_length(max_length))
+    }
+}
+
+impl Default for NewlineDelimitedCodec {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Decoder for NewlineDelimitedCodec {
+    type Item = Bytes;
+    type Error = BoxedFramingError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.0
+            .decode(src)
+            .map(|item| item.map(Into::into))
+            .map_err(Into::into)
+    }
+
+    fn decode_eof(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.0
+            .decode_eof(src)
+            .map(|item| item.map(Into::into))
+            .map_err(Into::into)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn decode_bytes_with_newlines() {
+        let mut input = BytesMut::from("foo\nbar\nbaz");
+        let mut decoder = NewlineDelimitedCodec::new();
+
+        assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "foo");
+        assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "bar");
+        assert_eq!(decoder.decode(&mut input).unwrap(), None);
+    }
+
+    #[test]
+    fn decode_bytes_with_newlines_trailing() {
+        let mut input = BytesMut::from("foo\nbar\nbaz\n");
+        let mut decoder = NewlineDelimitedCodec::new();
+
+        assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "foo");
+        assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "bar");
+        assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "baz");
+        assert_eq!(decoder.decode(&mut input).unwrap(), None);
+    }
+
+    #[test]
+    fn decode_bytes_with_newlines_and_max_length() {
+        let mut input = BytesMut::from("foo\nbarbara\nbaz\n");
+        let mut decoder = NewlineDelimitedCodec::new_with_max_length(3);
+
+        assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "foo");
+        assert!(decoder.decode(&mut input).is_err());
```

Comment:
This is a very good question. I agree that `max_length` leaves too much ambiguity for interpretation.

I'd suggest renaming it to `max_length_drop` and implement the behavior you proposed or rename it to `max_length_truncate`?

Reply:
I think it's a reasonable default to drop invalid frames instead of passing them on to a parser that's unlikely to succeed. We could add an option to truncate and pass, but I'd rather wait for someone to ask for that functionality before we go ahead and implement it.

Reply:
We'll move this to the PR that implements `newline_delimited` framer.